ir/src/main/scala/org/scalajs/core/ir/Trees.scala

```diff
@@ -436,6 +436,122 @@ object Trees {
+   *
+   *  {{{
+   *  @ScalaJSDefined
+   *  class Foo extends Bar
+   *  }}}
+   *
+   *  The node
+   *
+   *  {{{
+   *  JSBrackerSuperCall(ClassType(Foo), receiver, method, args)
+   *  }}}
+   *
+   *  which is printed as
+   *
+   *  {{{
+   *  receiver.Foo::super[method](...args)
+   *  }}}
+   *
+   *  has the following semantics:
+   *
+   *  {{{
+   *  Bar.prototype[method].call(receiver, ...args)
+   *  }}}
+   *
+   *  If this happens to be located in an instance method of `Foo`, *and*
+   *  `receiver` happens to be `This()`, this is equivalent to the ES6
+   *  statement
+   *
+   *  {{{
+   *  super[method](...args)
+   *  }}}
+   */
+  case class JSBracketSuperCall(cls: ClassType, receiver: Tree, method: Tree,
+      args: List[Tree])(implicit val pos: Position) extends Tree {
+    val tpe = AnyType
+  }
+
+  /** Super constructor call in the constructor of a Scala.js-defined JS class.
+   *
+   *  Exactly one such node must appear in the constructor of a
+   *  Scala.js-defined JS class, at the top-level (possibly as a direct child
+   *  of a top-level `Block`).
+   *
+   *  Statements before this node, as well as the `args`, cannot contain any
+   *  `This()` node. Statements after this node can use `This()`.
+   *
+   *  This node's semantics are to call the super constructor, then create
+   *  the fields of the object declared by the current class, initialized to
+   *  the zero of their type.
+   *
+   *  All in all, the shape of a constructor is therefore:
+   *
+   *  {{{
+   *  {
+   *    statementsNotUsingThis();
+   *    JSSuperConstructorCall(...argsNotUsingThis);
+   *    statementsThatMayUseThis()
+   *  }
+   *  }}}
+   *
+   *  which translates to something of the following shape:
+   *
+   *  {{{
+   *  {
+   *    statementsNotUsingThis();
+   *    super(...argsNotUsingThis);
+   *    this.privateField1 = 0;
+   *    this["publicField2"] = false;
+   *    statementsThatMayUseThis()
+   *  }
+   *  }}}
+   *
+   *  Any other use of this node is invalid.
+   */
+  case class JSSuperConstructorCall(args: List[Tree])(
+      implicit val pos: Position) extends Tree {
+    val tpe = NoType
+  }
+
+  /** Loads the JS constructor for a Scala.js-defined JS class.
+   *
+   *  `cls` must represent a Scala.js-defined JS class.
+   *
+   *  This is used typically to instantiate a Scala.js-defined JS class. Given
+   *  the class
+   *
+   *  {{{
+   *  @ScalaJSDefined
+   *  class Foo(x: Int) extends js.Object
+   *  }}}
+   *
+   *  The instantiation `new Foo(1)` would be represented as
+   *
+   *  {{{
+   *  JSNew(JSLoadConstructor(ClassType("Foo")), List(IntLiteral(1)))
+   *  }}}
+   *
+   *  This node is also useful to encode `o.isInstanceOf[Foo]`:
+   *
+   *  {{{
+   *  JSBinaryOp(instanceof, o, JSLoadConstructor(ClassType("Foo")))
+   *  }}}
+   *
+   *  The presence of this node makes the class `Foo` instantiable, and
+   *  therefore reachable.
+   */
+  case class JSLoadConstructor(cls: ClassType)(
```

Comment:
Actually, I will expand the semantics of this node to also be valid for a native JS class. This is necessary for native vs non-native to be an implementation detail. And that will make the "abuse" in desugaring a non-abuse.

And in that case, it might one day have to actually *load* stuff. We'll need some loading semantics when we have some support for `require`-style loading of native classes and objects.